File: turbo/apps/web/src/lib/e2b/scripts/log.ts
/**
 * Unified logging functions for agent scripts
 * Provides consistent log format across all sandbox scripts
 */
export const LOG_SCRIPT = `# Unified logging functions
# Format: [VM0][LEVEL] message

log_info() {
  echo "[VM0][INFO] $*" >&2
}

log_warn() {
  echo "[VM0][WARN] $*" >&2
}

log_error() {
  echo "[VM0][ERROR] $*" >&2
}

log_debug() {
  if [ "$VM0_DEBUG" = "1" ]; then
    echo "[VM0][DEBUG] $*" >&2
  fi
}
`;

File: crates/vm-common/src/log.rs
//! Logging utilities for VM scripts.

/// Get current timestamp in RFC3339 format with milliseconds.
pub fn timestamp() -> String {
    chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true)
}

/// Log an info message to stderr.
#[macro_export]
macro_rules! log_info {
    ($tag:expr, $($arg:tt)*) => {
        eprintln!("[{}] [INFO] [{}] {}", $crate::log::timestamp(), $tag, format!($($arg)*));
    };
}

/// Log a warning message to stderr.
#[macro_export]
macro_rules! log_warn {
    ($tag:expr, $($arg:tt)*) => {
        eprintln!("[{}] [WARN] [{}] {}", $crate::log::timestamp(), $tag, format!($($arg)*));
    };
}

/// Log an error message to stderr.
#[macro_export]
macro_rules! log_error {
    ($tag:expr, $($arg:tt)*) => {
        eprintln!("[{}] [ERROR] [{}] {}", $crate::log::timestamp(), $tag, format!($($arg)*));
    };
}
